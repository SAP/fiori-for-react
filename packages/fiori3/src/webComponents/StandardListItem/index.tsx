import { ListItemTypes } from '../../lib/ListItemTypes';
import UI5StandardListItem from '@ui5/webcomponents/dist/StandardListItem';
import React, { FC } from 'react';
import { withWebComponent, WithWebComponentPropTypes } from '../../internal/withWebComponent';

export interface StandardListItemPropTypes extends WithWebComponentPropTypes {
  selected?: boolean; // @generated
  type?: ListItemTypes; // @generated
  description?: string; // @generated
  icon?: string; // @generated
  iconEnd?: boolean; // @generated
  image?: string; // @generated
  children?: any[]; // @generated
}

const StandardListItem: FC<StandardListItemPropTypes> = withWebComponent<StandardListItemPropTypes>(
  UI5StandardListItem
);

StandardListItem.displayName = 'StandardListItem';

StandardListItem.defaultProps = {
  type: ListItemTypes.Active // @generated
};

export { StandardListItem };
