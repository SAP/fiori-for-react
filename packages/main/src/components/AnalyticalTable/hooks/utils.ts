export const makeTemplateColumns = (columns, resizedColumns) =>
  columns
    .map((column) => {
      if (resizedColumns[column.id]) {
        return `${resizedColumns[column.id]}px`;
      }
      return `minmax(${column.minWidth}px, ${column.width || '1fr'})`;
    })
    .join(' ');
