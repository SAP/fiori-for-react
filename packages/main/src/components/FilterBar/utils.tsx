import { cloneElement, ReactElement } from 'react';

export const setPropsOfChildren = (children, currentRef) => {
  return children.map((child) => {
    if (child.props.children.props[currentRef]?.children) {
      const childrenOfChild = child.props.children;
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-SELECT' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-MULTI-COMBOBOX'
      ) {
        const selectedIndices = Array.from(child.props.children.props[currentRef].children)
          .map((item, index) => (item._state.selected ? index : false))
          .filter((el) => el !== false);
        const options = child.props.children.props.children.map((item, index) => {
          if (selectedIndices.includes(index)) {
            return cloneElement(item, { selected: true });
          }
          return cloneElement(item, { selected: false });
        });
        return cloneElement(child, {
          children: { ...childrenOfChild, props: { ...childrenOfChild.props, children: options } }
        });
      }
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-INPUT' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-DATEPICKER'
      ) {
        return cloneElement(child, {
          children: {
            ...childrenOfChild,
            props: { ...childrenOfChild.props, value: child.props.children.props[currentRef]._state.value }
          }
        });
      }
      if (
        child.props.children.props[currentRef]?.tagName === 'UI5-SWITCH' ||
        child.props.children.props[currentRef]?.tagName === 'UI5-CHECKBOX'
      ) {
        return cloneElement(child, {
          children: {
            ...childrenOfChild,
            props: { ...childrenOfChild.props, checked: child.props.children.props[currentRef]._state.checked }
          }
        });
      }
      return child;
    }
    return child;
  });
};

export const addRef = (children, refsArray, currentRef) =>
  children.map((child) => {
    const childRef = refsArray.filter((item) => item.key === child.key);
    const childrenAttributes = child.props.children;
    const childrenProps = child.props.children.props;
    return cloneElement(child as ReactElement<any>, {
      children: { ...childrenAttributes, props: { ...childrenProps, [currentRef]: childRef[0]?.node } }
    });
  });

export const renderSearchWithValue = (searchElement, searchValue) => {
  return cloneElement(searchElement(), { value: searchValue });
};
