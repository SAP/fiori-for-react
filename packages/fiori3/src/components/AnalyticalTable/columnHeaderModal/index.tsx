import { withStyles } from '@fiori-for-react/styles';
import { Event } from '@fiori-for-react/utils';
import React, { Component, ReactNode } from 'react';
import { ClassProps } from '../../../interfaces/ClassProps';
import { JSSTheme } from '../../../interfaces/JSSTheme';
import { List } from '../../../lib/List';
import { ListItemTypes } from '../../../lib/ListItemTypes';
import { PlacementType } from '../../../lib/PlacementType';
import { PopoverHorizontalAlign } from '../../../lib/PopoverHorizontalAlign';
import { Popover } from '../../../lib/Popover';
import { StandardListItem } from '../../../lib/StandardListItem';
import { ColumnType } from '../columnHeader';

const styles = ({ parameters }: JSSTheme) => ({
  modalRoot: {
    border: 'none',
    borderRadius: '0.25rem',
    boxShadow: parameters.sapUiShadowLevel1,
    '&:focus': {
      outline: 'none'
    }
  }
});

export interface ColumnHeaderModalProperties {
  openBy: ReactNode;
  showSort?: boolean;
  showFilter?: boolean;
  sortAscending: (event: Event) => void;
  sortDescending: (event: Event) => void;
  FilterComponent: any;
  filter: any;
  column: ColumnType;
}

interface ColumnHeaderModalInternalProperties extends ColumnHeaderModalProperties, ClassProps {}

@withStyles(styles)
export class ColumnHeaderModal extends Component<ColumnHeaderModalProperties> {
  static defaultProps = {
    showSort: true,
    showFilter: false,
    filter: null
  };

  private popoverRef = null;

  private handlePopoverRef = (el) => {
    this.popoverRef = el;
  };

  private handleSort = (e) => {
    const sortType = e.getParameter('item').getAttribute('data-sort');
    if (sortType === 'asc') {
      this.props.sortAscending(Event.of(this, e.getOriginalEvent()));
    } else {
      this.props.sortDescending(Event.of(this, e.getOriginalEvent()));
    }
    this.popoverRef.close();
  };

  render() {
    const { showSort } = this.props as ColumnHeaderModalInternalProperties;

    return (
      <Popover
        openByStyle={{ flex: '100 0 auto', width: '100px' }}
        openBy={this.props.openBy}
        hideHeader
        hideArrow
        horizontalAlign={PopoverHorizontalAlign.Left}
        placementType={PlacementType.Bottom}
        innerComponentRef={this.handlePopoverRef}
      >
        <List onItemPress={this.handleSort}>
          {showSort && (
            <StandardListItem type={ListItemTypes.Active} icon={'sap-icon://sort-ascending'} data-sort={'asc'}>
              Sort Ascending
            </StandardListItem>
          )}
          {showSort && (
            <StandardListItem type={ListItemTypes.Active} icon={'sap-icon://sort-descending'} data-sort={'desc'}>
              Sort Descending
            </StandardListItem>
          )}
        </List>
      </Popover>
    );
  }
}
