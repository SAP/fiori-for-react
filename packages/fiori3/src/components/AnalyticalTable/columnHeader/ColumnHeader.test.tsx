import React from 'react';
import { ColumnHeader } from './index';
import { mountThemedComponent } from '@shared/tests/utils';
import { ColumnHeaderModal } from '../columnHeaderModal/index';
import { StandardListItem } from '@lib/StandardListItem';

const column = {
  id: 'column1',
  show: true,
  sortable: true
};

const sortedWithColumn = [
  // is usually provided by react-table
  {
    id: 'column1',
    desc: true
  }
];

const filteredWithColumn = [
  // is usually provided by react-table
  {
    id: 'column1'
  }
];

describe('Table/ColumnHeader', () => {
  test('With Props', () => {
    // @ts-ignore
    const wrapper = mountThemedComponent(<ColumnHeader column={column} sorted={[]} filtered={[]} />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Render Sorted Column', () => {
    const wrapper = mountThemedComponent(
      // @ts-ignore
      <ColumnHeader column={column} sorted={sortedWithColumn} filtered={filteredWithColumn} />
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('invisible Column', () => {
    const col = {
      ...column,
      show: false
    };
    // @ts-ignore
    const wrapper = mountThemedComponent(<ColumnHeader column={col} sorted={[]} filtered={[]} />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Sort ascending', () => {
    const dummyToggleSort = () => null; // is usually being provided by react-table
    const wrapper = mountThemedComponent(
      // @ts-ignore
      <ColumnHeader column={column} toggleSort={dummyToggleSort} sortable sorted={[]} filtered={[]} />
    );
    wrapper.simulate('click');
    wrapper.update();
    const items = wrapper.find(ColumnHeaderModal).find(StandardListItem);
    items.at(0).simulate('click');
    wrapper.update();
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Sort descending', () => {
    const dummyToggleSort = () => null; // is usually being provided by react-table
    const wrapper = mountThemedComponent(
      // @ts-ignore
      <ColumnHeader column={column} toggleSort={dummyToggleSort} sortable sorted={[]} filtered={[]} />
    );
    wrapper.simulate('click');
    wrapper.update();
    const items = wrapper.find(ColumnHeaderModal).find(StandardListItem);
    items.at(1).simulate('click');
    wrapper.update();
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Sort when already sorted - ascending', () => {
    const dummyToggleSort = () => null; // is usually being provided by react-table
    const wrapper = mountThemedComponent(
      // @ts-ignore
      <ColumnHeader column={column} toggleSort={dummyToggleSort} sortable sorted={sortedWithColumn} filtered={[]} />
    );
    wrapper.simulate('click');
    wrapper.update();
    const items = wrapper.find(ColumnHeaderModal).find(StandardListItem);
    items.at(0).simulate('click');
    wrapper.update();
    expect(wrapper.render()).toMatchSnapshot();
  });

  test('Sort when already sorted - descending', () => {
    const dummyToggleSort = () => null; // is usually being provided by react-table
    const wrapper = mountThemedComponent(
      // @ts-ignore
      <ColumnHeader column={column} toggleSort={dummyToggleSort} sortable sorted={sortedWithColumn} filtered={[]} />
    );
    wrapper.simulate('click');
    wrapper.update();
    const items = wrapper.find(ColumnHeaderModal).find(StandardListItem);
    items.at(1).simulate('click');
    wrapper.update();
    expect(wrapper.render()).toMatchSnapshot();
  });
});
