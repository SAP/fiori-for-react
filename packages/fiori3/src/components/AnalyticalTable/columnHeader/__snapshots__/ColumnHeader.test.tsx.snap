// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table/ColumnHeader Render Sorted Column 1`] = `
<div
  class="rt-th ColumnHeader-header- ColumnHeader-header-"
>
  <div
    class="ColumnHeader-iconContainer-"
  >
    <span
      aria-label="filter"
      class="Icon-icon- Icon-icon-"
      data-icon-content=""
      role="presentation"
      style="font-family: SAP-icons; font-size: 0.75rem;"
      tabindex="-1"
    />
    <span
      aria-label="sort-descending"
      class="Icon-icon- Icon-icon-"
      data-icon-content=""
      role="presentation"
      style="font-family: SAP-icons; font-size: 0.75rem;"
      tabindex="-1"
    />
  </div>
</div>
`;

exports[`Table/ColumnHeader Sort ascending 1`] = `
<div
  class="rt-th ColumnHeader-header- ColumnHeader-header-"
>
  <div
    class="ColumnHeader-iconContainer-"
  />
</div>
`;

exports[`Table/ColumnHeader Sort descending 1`] = `
<div
  class="rt-th ColumnHeader-header- ColumnHeader-header-"
>
  <div
    class="ColumnHeader-iconContainer-"
  />
</div>
`;

exports[`Table/ColumnHeader Sort when already sorted - ascending 1`] = `
<div
  class="rt-th ColumnHeader-header- ColumnHeader-header-"
>
  <div
    class="ColumnHeader-iconContainer-"
  >
    <span
      aria-label="sort-descending"
      class="Icon-icon- Icon-icon-"
      data-icon-content=""
      role="presentation"
      style="font-family: SAP-icons; font-size: 0.75rem;"
      tabindex="-1"
    />
  </div>
</div>
`;

exports[`Table/ColumnHeader Sort when already sorted - descending 1`] = `
<div
  class="rt-th ColumnHeader-header- ColumnHeader-header-"
>
  <div
    class="ColumnHeader-iconContainer-"
  >
    <span
      aria-label="sort-descending"
      class="Icon-icon- Icon-icon-"
      data-icon-content=""
      role="presentation"
      style="font-family: SAP-icons; font-size: 0.75rem;"
      tabindex="-1"
    />
  </div>
</div>
`;

exports[`Table/ColumnHeader With Props 1`] = `
<div
  class="rt-th ColumnHeader-header- ColumnHeader-header-"
>
  <div
    class="ColumnHeader-iconContainer-"
  />
</div>
`;

exports[`Table/ColumnHeader invisible Column 1`] = `
<div
  class="rt-th ColumnHeader-header- ColumnHeader-header- -hidden"
>
  <div
    class="ColumnHeader-iconContainer-"
  />
</div>
`;
